### PR TITLE
[pose-detection]Implemented low pass filter and visibility low pass filter.

### PR DESCRIPTION
**Pull request by lina128**
Changes in this PR:
1. Implemented Mediapipe's generic low pass filter class in JS.
2. Implemented the visibilitySmoothingCalculator in Blazepose pipeline.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/617)

### CODE REVIEW

File: pose-detection/src/blazepose/calculators/visibility_smoothing.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+export class LowPassVisibilityFilter {
+  private alpha: number;
+  private visibilityFilters: LowPassFilter[];
+
+  constructor(config: VisibilitySmoothingCalculatorConfig) {
+    this.alpha = config.alpha;
+  };
+
+  apply(landmarks: Keypoint[]) {
+    if (this.visibilityFilters == null) {
```
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose/calculators/visibility_smoothing.ts:34]: Just to confirm, when `visibilityFilters` is initialized it is `null` or empty array?

**Review by ivan-grishchenko — APPROVED** [review]

**Review by lina128 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/617)* status: 0 of 1 approvals obtained (waiting on @bazarevsky, @ivan-grishchenko, @mhays-google, and @pyu10055)

___
*[pose-detection/src/blazepose/calculators/visibility_smoothing.ts, line 34 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWf2sPJtoiTi_vHyyPe-r1-34:-MWf7LCw2ZWTGFB1ajfA:bymthun) ([raw file](https://github.com/tensorflow/tfjs-models/blob/73aef78ab21e58c9c03b4b85f775784ceffc5e18/pose-detection/src/blazepose/calculators/visibility_smoothing.ts#L34)):*
<details><summary><i>Previously, ivan-grishchenko wrote…</i></summary><blockquote>

Just to confirm, when `visibilityFilters` is initialized it is `null` or empty array?
</blockquote></details>

It will be `undefined`,  `undefined == null` will be truthy.

**Review by pyu10055 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/617)* status: 0 of 1 approvals obtained (waiting on @bazarevsky, @ivan-grishchenko, @lina128, and @mhays-google)

___
*[pose-detection/src/blazepose/constants.ts, line 48 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWfyc4s311EQA30PGx-:-MWfyc4s311EQA30PGx0:bm43qws) ([raw file](https://github.com/tensorflow/tfjs-models/blob/f20fb46dbb649e36067f372c1793f1560b63b33a/pose-detection/src/blazepose/constants.ts#L48)):*
> ```XML
>   maxPoses: 1,
>   flipHorizontal: false,
>   enableSmoothing: false
> ```

should the smoothing flag be enabled by default?
___
*[pose-detection/src/blazepose/calculators/visibility_smoothing.ts, line 33 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWfzA9l-MoM8WrX3QqP:-MWfzA9l-MoM8WrX3QqQ:b2q75h9) ([raw file](https://github.com/tensorflow/tfjs-models/blob/f20fb46dbb649e36067f372c1793f1560b63b33a/pose-detection/src/blazepose/calculators/visibility_smoothing.ts#L33)):*
> ```XML
>   };
> 
>   apply(landmarks: Keypoint[]) {
> ```

is landmarks length always the same?
___
*[pose-detection/src/calculators/low_pass_filter.ts, line 27 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWfzx7w9peFTXM886oj:-MWfzx7xBVnaTgFl1ipr:b-hhqtdz) ([raw file](https://github.com/tensorflow/tfjs-models/blob/f20fb46dbb649e36067f372c1793f1560b63b33a/pose-detection/src/calculators/low_pass_filter.ts#L27)):*
> ```XML
> // ref:
> // https://github.com/google/mediapipe/blob/master/mediapipe/util/filtering/low_pass_filter.cc
> export class LowPassFilter {
> ```

should low pass Filter be FPS sensitive? or it should have an expiration limit. If the inference is paused, your inputs are not continuous, will the filter catch up quickly?

**Review by lina128 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/617)* status: 0 of 1 approvals obtained (waiting on @bazarevsky, @ivan-grishchenko, @mhays-google, and @pyu10055)

___
*[pose-detection/src/blazepose/constants.ts, line 48 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWfyc4s311EQA30PGx-:-MWg-e619J6woiNmwJMx:b-bl3pas) ([raw file](https://github.com/tensorflow/tfjs-models/blob/f20fb46dbb649e36067f372c1793f1560b63b33a/pose-detection/src/blazepose/constants.ts#L48)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

should the smoothing flag be enabled by default?
</blockquote></details>

I think default values are generally false, in other words, they are opt-in options. Same reason for why setting flipHorizontal to false by default.
___
*[pose-detection/src/blazepose/calculators/visibility_smoothing.ts, line 33 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWfzA9l-MoM8WrX3QqP:-MWg1a7e3uYK9gUEXpCZ:bylijwk) ([raw file](https://github.com/tensorflow/tfjs-models/blob/f20fb46dbb649e36067f372c1793f1560b63b33a/pose-detection/src/blazepose/calculators/visibility_smoothing.ts#L33)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

is landmarks length always the same?
</blockquote></details>

I think so, because number of landmarks are determined by model, which is determined when the detector is created. But this is good suggestion for safer code, so I added a length check and initiate new filters if length doesn't match. For reference, the source code threw error if length doesn't match.
___
*[pose-detection/src/calculators/low_pass_filter.ts, line 27 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWfzx7w9peFTXM886oj:-MWg2Po_7vY3WJltMN5S:b-yqm2ng) ([raw file](https://github.com/tensorflow/tfjs-models/blob/f20fb46dbb649e36067f372c1793f1560b63b33a/pose-detection/src/calculators/low_pass_filter.ts#L27)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

should low pass Filter be FPS sensitive? or it should have an expiration limit. If the inference is paused, your inputs are not continuous, will the filter catch up quickly?
</blockquote></details>

Hmm, that's a good question. @ivan-grishchenko What do you think? Should the visibility filter be sensitive to how much time have elapsed between the current value and previous value?

**Review by lina128 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/617)* status: 0 of 1 approvals obtained (waiting on @bazarevsky, @ivan-grishchenko, @mhays-google, and @pyu10055)

___
*[pose-detection/src/blazepose/constants.ts, line 48 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWfyc4s311EQA30PGx-:-MWgTrZwCabmz0phzcTp:b5dtxpa) ([raw file](https://github.com/tensorflow/tfjs-models/blob/f20fb46dbb649e36067f372c1793f1560b63b33a/pose-detection/src/blazepose/constants.ts#L48)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

I think default values are generally false, in other words, they are opt-in options. Same reason for why setting flipHorizontal to false by default.
</blockquote></details>

Given that MediaPipe set it to true by default, we'll also set this to true.

**Review by pyu10055 — APPROVED** [review]
Thanks I will LGTMed, and if we can figure out the filter expiration, it can be added later.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-:-MWgg0iL-hQLpEgjtGAC:bekd74s)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @bazarevsky, @ivan-grishchenko, @mhays-google, and @pyu10055)

___
*[pose-detection/src/blazepose/constants.ts, line 48 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWfyc4s311EQA30PGx-:-MWg5dT67EPlrPTnGxn7:bcnwsyr) ([raw file](https://github.com/tensorflow/tfjs-models/blob/f20fb46dbb649e36067f372c1793f1560b63b33a/pose-detection/src/blazepose/constants.ts#L48)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

I think default values are generally false, in other words, they are opt-in options. Same reason for why setting flipHorizontal to false by default.
</blockquote></details>

Is the flag enabled by default on MP side? I think this flag is critical for video application, otherwise there will be too much jittering

**Review by pyu10055 — COMMENTED** [review]
> Reviewed 4 of 36 files at r1, 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/617)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @bazarevsky, @ivan-grishchenko, @mhays-google, and @pyu10055)

**Review by ivan-grishchenko — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/617)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @bazarevsky, @ivan-grishchenko, @mhays-google, and @pyu10055)

___
*[pose-detection/src/blazepose/calculators/visibility_smoothing.ts, line 33 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWfzA9l-MoM8WrX3QqP:-MWyx8UW8lWPLZz7T0Ow:bbj2ln0) ([raw file](https://github.com/tensorflow/tfjs-models/blob/f20fb46dbb649e36067f372c1793f1560b63b33a/pose-detection/src/blazepose/calculators/visibility_smoothing.ts#L33)):*
> number
FYI: In MediaPipe we initialize filters when they are null (on first frame) and throw error it later we get different number of landmarks (it shouldn't happen)
___
*[pose-detection/src/calculators/low_pass_filter.ts, line 27 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWfzx7w9peFTXM886oj:-MWyya3VEy2bsxJZbs8_:b-vjr56r) ([raw file](https://github.com/tensorflow/tfjs-models/blob/f20fb46dbb649e36067f372c1793f1560b63b33a/pose-detection/src/calculators/low_pass_filter.ts#L27)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

Hmm, that's a good question. @ivan-grishchenko What do you think? Should the visibility filter be sensitive to how much time have elapsed between the current value and previous value?
</blockquote></details>

It's a nice thing to add in future to support floating FPS (i.e. weight every next value depending on time elapsed since the last event).
But while it's much more important for landmarks coordinates (and is supported in relative velocity filter), visibility is not that sensitive to floating FPS.
Regarding the case when inference is paused, we don't have it in MediaPipe, but when we have no landmarks prediction on some frame, we reset filters. You should do the same when inference is stopped.

**Review by lina128 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/617)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @bazarevsky, @ivan-grishchenko, @mhays-google, and @pyu10055)

___
*[pose-detection/src/calculators/low_pass_filter.ts, line 27 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWfzx7w9peFTXM886oj:-MWz-SS83EoYn-K-cmfR:bgbg71q) ([raw file](https://github.com/tensorflow/tfjs-models/blob/f20fb46dbb649e36067f372c1793f1560b63b33a/pose-detection/src/calculators/low_pass_filter.ts#L27)):*
<details><summary><i>Previously, ivan-grishchenko wrote…</i></summary><blockquote>

It's a nice thing to add in future to support floating FPS (i.e. weight every next value depending on time elapsed since the last event).
But while it's much more important for landmarks coordinates (and is supported in relative velocity filter), visibility is not that sensitive to floating FPS.
Regarding the case when inference is paused, we don't have it in MediaPipe, but when we have no landmarks prediction on some frame, we reset filters. You should do the same when inference is stopped.
</blockquote></details>

Got it. Thanks for the detailed explanation. The landmarks coordinate filter will have the reset when no prediction is detected.

**Review by bazarevsky — APPROVED** [review]

**Review by mhays-google — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/617)* status: :shipit: complete! 3 of 1 approvals obtained (waiting on @ivan-grishchenko, @lina128, and @pyu10055)

___
*[pose-detection/src/blazepose/calculators/visibility_smoothing.ts, line 33 at r4](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWz7oZVAN73JUHXCV4W:-MWz7oZVAN73JUHXCV4X:bzfsyv2) ([raw file](https://github.com/tensorflow/tfjs-models/blob/f0103842459ed6cc515ecb7f6f266a8be8d79e38/pose-detection/src/blazepose/calculators/visibility_smoothing.ts#L33)):*
> ```
>  
> ```
return type?
___
*[pose-detection/src/calculators/low_pass_filter.ts, line 32 at r4](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWz8-7T4dDdLUr22D-z:-MWz8-7T4dDdLUr22D0-:b-y1gkxh) ([raw file](https://github.com/tensorflow/tfjs-models/blob/f0103842459ed6cc515ecb7f6f266a8be8d79e38/pose-detection/src/calculators/low_pass_filter.ts#L32)):*
> ```
> his.initialized = false;
> ```
nit: You can do this right at the member variable decl.

**Review by lina128 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/617)* status: :shipit: complete! 3 of 1 approvals obtained (waiting on @ivan-grishchenko, @mhays-google, and @pyu10055)

___
*[pose-detection/src/blazepose/calculators/visibility_smoothing.ts, line 33 at r4](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWz7oZVAN73JUHXCV4W:-MWz8vVL2w1NXFUjKovI:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/f0103842459ed6cc515ecb7f6f266a8be8d79e38/pose-detection/src/blazepose/calculators/visibility_smoothing.ts#L33)):*
<details><summary><i>Previously, mhays-google (Michael Hays) wrote…</i></summary><blockquote>

> ```
>  
> ```
return type?
</blockquote></details>

Done.
___
*[pose-detection/src/calculators/low_pass_filter.ts, line 32 at r4](https://reviewable.io/reviews/tensorflow/tfjs-models/617#-MWz8-7T4dDdLUr22D-z:-MWz934z-7fKaPioEkF3:bs7wumb) ([raw file](https://github.com/tensorflow/tfjs-models/blob/f0103842459ed6cc515ecb7f6f266a8be8d79e38/pose-detection/src/calculators/low_pass_filter.ts#L32)):*
<details><summary><i>Previously, mhays-google (Michael Hays) wrote…</i></summary><blockquote>

> ```
> his.initialized = false;
> ```
nit: You can do this right at the member variable decl.
</blockquote></details>

Done.

**Pull request merged**